### PR TITLE
Move startup validation to the options assembly

### LINKED ISSUE #84347: [API Proposal]: Decouple ValidateOnStart from Hosting
Usage of [ValidateOnStart()](https://learn.microsoft.com/dotnet/api/microsoft.extensions.dependencyinjection.optionsbuilderextensions.validateonstart?view=dotnet-plat-ext-7.0) today is located in the `Microsoft.Extensions.Hosting` assembly\package. This is problematic as it greatly inflates dependency tree of packages that need to validate their options eagerly when the hosting package is not otherwise needed. The `ValidateOnStart()` functionality should be moved to the `Microsoft.Extensions.Options` package.

In order for a host to call a "Validate()" method, a new public interface needs to be added, called `IValidateOnStart` here. This interface will be injected by the `ValidateOnStart()` functionality and then obtained by a host (for the default host, in the  `Microsoft.Extensions.Hosting` assembly) with a call to `GetService<IValidateOnStart>()` in order to trigger the validation during startup.

The new interface located in the `Microsoft.Extensions.Options` assembly:
```diff
namespace Microsoft.Extensions.Options
{
+    public interface IValidateOnStart
+    {
+        void Validate();
+   }
}
```

The existing `ValidateOnStart()` extension method is located on the `Microsoft.Extensions.DependencyInjection.OptionsBuilderExtensions` extension class in the `Microsoft.Extensions.Hosting` assembly. The `Microsoft.Extensions.Hosting` assembly needs to forward this to the `Microsoft.Extensions.Options` assembly:
```diff
+[assembly: System.Runtime.CompilerServices.TypeForwardedTo(
+    typeof(Microsoft.Extensions.DependencyInjection.OptionsBuilderExtensions))]
```

### Background and motivation

Usage of `ValidateOnStart` today is tightly coupled with Microsoft.Extensions.Hosting package. The call registers both `ValidationHostedService` and instructs `ValidateOnStart` infrastructure to validate `T`-type upon startup. This is problematic as it greatly inflates dependency tree of packages that need to validate their options eagerly.

Following measures should be considered:
- Move `ValidateOnStart` to Microsoft.Extensions.Options package. This decouple registration from implementation. Code that expresses the desire for eager validation doesn't involve registration of ValidationHostedService in the same call. The new API enables a cleaner separation of concern.
- Incorporate `ValidationHostedService` into `Host` bootstrapping as a shared option infrastructure. It becomes part of default `Host` and doesn't need to be registered with every `ValidateOnStart` call.

### API Proposal

```csharp
namespace Microsoft.Extensions.Options
{
    public static class OptionsBuilderExtensions
    {
        public static OptionsBuilder<TOptions> ValidateOnStart<TOptions>(this OptionsBuilder<TOptions> builder) 
            where TOptions : class;
    }
```

(Shape of the API remains the same)

### API Usage

(Usage of the API remains the same.)

### Alternative Designs

_No response_

### Risks

Moving types around.

### PR DESCRIPTION
**Pull request by steveharter**
Fixes https://github.com/dotnet/runtime/issues/84347

### CODE REVIEW

**Review by buyaa-n — APPROVED** [review]
Left a few NIT, overall LGTM

File: src/libraries/Microsoft.Extensions.Options/ref/Microsoft.Extensions.Options.cs [hunk 1]
```diff
@@ -135,6 +139,10 @@ public partial interface IPostConfigureOptions<in TOptions> where TOptions : cla
     {
         void PostConfigure(string? name, TOptions options);
     }
+    public partial interface IStartupValidator
```
[the remark this answers is not in the document]
Comment (steveharter) [reply src/libraries/Microsoft.Extensions.Options/ref/Microsoft.Extensions.Options.cs:142]: It's in the `/ref/` file which the API gen tool generates the `partial` modifier.

File: src/libraries/Microsoft.Extensions.Options/ref/Microsoft.Extensions.Options.csproj [hunk 2]
```diff
@@ -23,5 +23,6 @@
 
   <ItemGroup Condition="'$(TargetFrameworkIdentifier)' == '.NETFramework'">
     <Reference Include="System.ComponentModel.DataAnnotations" />
+    <PackageReference Include="System.ValueTuple" Version="$(SystemValueTupleVersion)" />
```
[the remark this answers is not in the document]
Comment (steveharter) [reply src/libraries/Microsoft.Extensions.Options/ref/Microsoft.Extensions.Options.csproj:26]: Yep thanks

File: src/libraries/Microsoft.Extensions.Options/src/IStartupValidator.cs [hunk 3]
```diff
@@ -0,0 +1,14 @@
...
+
+namespace Microsoft.Extensions.Options
+{
+    /// <summary>
+    /// Interface used by hosts to validate options during startup.
+    /// </summary>
+    /// <seealso cref="DependencyInjection.OptionsBuilderExtensions.ValidateOnStart{TOptions}(OptionsBuilder{TOptions})"/>
+    public interface IStartupValidator
+    {
+        void Validate();
```
[the remark this answers is not in the document]
Comment (steveharter) [reply src/libraries/Microsoft.Extensions.Options/src/IStartupValidator.cs:16]: Yep thanks

**Pull request merged**